File: 86-partition-list/86-partition-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* p1 = new ListNode(0);
        ListNode* p2 = new ListNode(0);
        ListNode* big = p1;
        ListNode* small  = p2;
        while(head != NULL)
        {
            ListNode* l = new ListNode(head->val);
            if(head->val >= x)
            {
                big->next = l;
                big = big->next;
            }
            else
            {
                small->next = l;
                small = small->next;
            }
            head = head->next;
        }
        
        small->next = p1->next;
        return p2->next;
        
        
    }
};

File: 86-partition-list/86-partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode h, int x) {
        ListNode a1 = new ListNode(0);
        ListNode a2 = new ListNode(0);
        ListNode h1 = a1;
        ListNode h2 = a2;
        while(h!=null)
        {
            if(h.val < x)
            {
               a1.next = h;
               h = h.next;
               a1 = a1.next;
               a1.next = null;
            }
            else
            {
                a2.next = h;
               h = h.next;
               a2 = a2.next;
               a2.next = null;
            }
        }
        
        a1.next = h2.next;
        return h1.next;
        
    }
}